File: authorization/src/main/kotlin/com/example/authorization/filter/JWTTokenValidatorFilter.java
package com.example.authorization.filter;

import io.jsonwebtoken.*;
import jakarta.servlet.*;
import jakarta.servlet.http.*;
import lombok.*;
import org.jetbrains.annotations.*;
import org.springframework.security.authentication.*;
import org.springframework.security.core.*;
import org.springframework.security.core.authority.*;
import org.springframework.security.core.context.*;
import org.springframework.web.filter.*;

import java.io.*;
import java.nio.charset.*;
import java.util.*;

@RequiredArgsConstructor
public class JWTTokenValidatorFilter extends OncePerRequestFilter {

    private final String jwtHeader;
    private final String jwtKey;

    @Override
    protected void doFilterInternal(HttpServletRequest request, @NotNull HttpServletResponse response,
                                    @NotNull FilterChain chain) throws ServletException, IOException {
        String jwt = request.getHeader(jwtHeader);
        if (jwt != null) {
            Claims claims = parseJwt(jwt);
            String username = String.valueOf(claims.get("username"));
            String authorities = String.valueOf(claims.get("authorities"));
            Authentication auth = createAuthentication(username, authorities);
            SecurityContextHolder.getContext().setAuthentication(auth);
        }
        chain.doFilter(request, response);
    }

    private Claims parseJwt(String jwt) {
        try {
            return Jwts.parserBuilder()
                    .setSigningKey(jwtKey.getBytes(StandardCharsets.UTF_8))
                    .build()
                    .parseClaimsJws(jwt)
                    .getBody();
        } catch (JwtException e) {
            throw new BadCredentialsException("Invalid Token received!", e);
        }
    }

    private Authentication createAuthentication(String username, String authorities) {
        List<GrantedAuthority> authorityList = AuthorityUtils.commaSeparatedStringToAuthorityList(authorities);
        return new UsernamePasswordAuthenticationToken(username, null, authorityList);
    }
}

File: authorization/src/main/kotlin/com/example/authorization/filter/JwtTokenValidatorFilter.kt
package com.example.authorization.filter

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import lombok.RequiredArgsConstructor
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.AuthorityUtils
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.filter.OncePerRequestFilter
import java.nio.charset.StandardCharsets

@RequiredArgsConstructor
class JWTTokenValidatorFilter(
    private val jwtHeader: String,
    private val jwtKey: String
) : OncePerRequestFilter() {

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        chain: FilterChain
    ) {
        val jwt = request.getHeader(jwtHeader)
        if (jwt != null) {
            val claims = parseJwt(jwt)
            val username = claims["username"].toString()
            val authorities = claims["authorities"].toString()
            val auth = createAuthentication(username, authorities)
            SecurityContextHolder.getContext().authentication = auth
        }
        chain.doFilter(request, response)
    }

    private fun parseJwt(jwt: String): Claims =
        Jwts.parserBuilder()
            .setSigningKey(jwtKey.toByteArray(StandardCharsets.UTF_8))
            .build()
            .parseClaimsJws(jwt)
            .body

    private fun createAuthentication(username: String, authorities: String): Authentication {
        val authorityList = AuthorityUtils.commaSeparatedStringToAuthorityList(authorities)
        return UsernamePasswordAuthenticationToken(username, null, authorityList)
    }
}
